src/Tools/DependencyManagement/ObjectContainer.php

```diff
@@ -0,0 +1,169 @@
+ *   - If the above is not possible, you can add an argument of type `ObjectContainer` to the class constructor,
+ *     and then use the `get` method of the supplied instance whenever needed.
+ *   - For code outside classes (inside standalone functions) or in static class methods, use
+ *    `ObjectContainer::get_instance_of`.
+ *
+ * Note: autowiring means the ability to automatically resolve a class name to an instance of that class via
+ *       `get(TheClass::class)` or by adding a typed argument to class constructors.
+ */
+final class ObjectContainer {
+
+	/**
+	 *  The singleton instance of ObjectContainer.
+	 *
+	 * @var ObjectContainer
+	 */
+	private static $instance = null;
+
+	/**
+	 *  The underlying dependency injection engine used.
+	 *
+	 * @var Container
+	 */
+	private $container;
+
+	/**
+	 *  The underlying reflection delegate used for autowiring.
+	 *
+	 * @var ReflectionContainer
+	 */
+	private $reflection_container;
+
+	/**
+	 *  Get the singleton instance of ObjectContainer.
+	 *
+	 * @return ObjectContainer
+	 */
+	public static function instance() {
+		return self::$instance;
+	}
+
+	/**
+	 * Initialize the dependency injection engine and do all the necessary registrations.
+	 */
+	public static function init() {
+		global $wp_actions;
+
+		if ( ! function_exists( 'did_action' ) || did_action( 'plugins_loaded' ) ) {
+			self::_init();
+		} else {
+			// TODO: This is a hack to overcome the inability to use the autoloader before plugins have been loaded, remove after https://github.com/Automattic/jetpack/pull/15106 is merged.
+			// phpcs:ignore WordPress.WP.GlobalVariablesOverride.Prohibited
+			$wp_actions['plugins_loaded'] = 1;
+			self::_init();
+			unset( $wp_actions['plugins_loaded'] );
+		}
+	}
+
+	/**
+	 * Initialize the dependency injection engine and do all the necessary registrations.
+	 */
+	public static function _init() {
+		// Instantiate the underlying container and create the singleton instance of ourselves.
+		$container      = new Container();
+		self::$instance = new self( $container );
+
+		// Register our singleton instance so we're able to resolve ourselves.
+		$container->add( __CLASS__, self::$instance );
+
+		// Register the reflection container to allow autowiring.
+		self::$instance->reflection_container = new WooReflectionContainer();
+		$container->delegate( self::$instance->reflection_container );
+
+		// Perform any required manual class and service provider registration.
+		self::$instance->register_classes();
+		self::$instance->register_service_providers();
+	}
+
+	/**
+	 * Class constructor.
+	 *
+	 * @param Container $container The instance of League\Container\Container to use for dependency injection.
+	 */
+	public function __construct( Container $container ) {
+		$this->container = $container;
+	}
+
+	/**
+	 * Register class resolutions for which default autowiring is not appropriate/enough.
+	 */
+	private function register_classes() {
```

Comment:
Yeah, why not. `Misc` or `General` would however be clearer than `Woo` as a prefix.

Reply:
I've tried but there's a problem. I need to put the list of shared classes in the `$provides` property of the service provider, otherwise it will never be executed. But since I'm registering the classes in the reflection container instead of the main container, I get a `Service provider lied about providing (class name) service` exception since only the main container is checked when dealing with service providers.

Reply:
This might just go back to us using the ReflectionContainer wrong here. Let's revisit with the ThemeSupport work as a good way to register those kinds of services perhaps?